code/tracing/TraceEventWriter.cpp

```diff
@@ -0,0 +1,111 @@
+#include "tracing/TraceEventWriter.h"
+
+namespace
+{
+using namespace tracing;
+
+const char* getTypeStr(tracing::EventType type) {
+	switch(type) {
+		case tracing::EventType::Complete:
+			return "X";
+		case tracing::EventType::Begin:
+			return "B";
+		case tracing::EventType::End:
+			return "E";
+		case EventType::AsyncBegin:
+			return "b";
+		case EventType::AsyncStep:
+			return "n";
+		case EventType::AsyncEnd:
+			return "e";
+		default: 
+			Assertion(false, "Invalid enum value!");
+			return "";
+	}
+}
+
+void writeTime(std::ofstream& out, std::uint64_t time) {
+	// Save stream state
+	auto flags = out.flags();
+	out << std::fixed;
+
+	out << (time / 1000.);
+
+	// and now restore it
+	out.flags(flags);
+}
+
+void writeCompleteEvent(std::ofstream& out_str, const trace_event* evt) {
+	Assertion(evt->type == EventType::Complete, "Event must be a complete event!");
+
+	out_str << ",\"dur\":";
+	writeTime(out_str, evt->duration);
+}
+}
+
+namespace tracing
+{
+
+TraceEventWriter::TraceEventWriter() : _out("tracing/trace.json") {
+	_out << "[";
+}
+
+TraceEventWriter::~TraceEventWriter() {
+	_out << "]\n";
+	_out.close();
+}
+
+void TraceEventWriter::processEvent(const trace_event* event) {
+	if (event->type == EventType::Complete) {
+		if (event->duration < 1000) {
+			// Discard events that are less than a millisecond long
+			return;
+		}
+	}
+
+	if (!_first_line) {
+		_out << ",";
+	}
+	_out << "\n{\"tid\": " << event->tid << ",\"ts\":";
+
+	writeTime(_out, event->timestamp);
+
+	_out << ",\"pid\":";
+	if (event->pid == GPU_PID) {
+		_out << "\"GPU\"";
```

Comment:
The code uses the timestamps provided by OpenGL so it doesn't matter if there are multiple GPUs since that would be hidden behind the OpenGL API.